### PR TITLE
OTC-856: Duplicate product mutation

### PR DESCRIPTION
**Pull request by wzglinieckisoldevelo**
Ticket: https://openimis.atlassian.net/browse/OTC-856

Changes:
- New mutation for duplication of product.

### CODE REVIEW

**Comment by delcroip** [comment]
is it still a draft ?

File: product/gql_mutations.py [hunk 1]
```diff
@@ -353,6 +356,70 @@ def async_mutate(cls, user, **data):
...
+
+        duplicate_items = True if 'items' not in data else False
+        duplicate_services = True if 'services' not in data else False
+
+        new_product = create_or_update_product(user, data, is_duplicate=True)
+
+        if duplicate_items:
+            new_product_items = ProductItem.objects.filter(product=Product.objects.get(uuid=current_uuid,
+                                                           validity_to__isnull=True))
+            for item in new_product_items:
```
Comment (delcroip) [inline product/gql_mutations.py:407]: are you stealing the item from the old product ? I don't see where you duplicate the item and services instance (create new item and then just copy the required fields)

Comment (wzglinieckisoldevelo) [reply product/gql_mutations.py:407]: Yes, it is the only feasible way. ProductItem contains information about connection between Item and Product. I would have to query this either way and then insert the same info basically. This code copies those elements, old data does not get removed. See:
https://docs.djangoproject.com/en/4.1/topics/db/queries/#copying-model-instances

Comment (delcroip) [reply product/gql_mutations.py:407]: @Thanks that make sense, I was afraid the "save_history" function was called upon `save()`

File: product/gql_mutations.py [hunk 2]
```diff
@@ -353,6 +356,70 @@ def async_mutate(cls, user, **data):
...
+                item.pk = None
```
Comment (delcroip) [inline product/gql_mutations.py:408]:
```suggestion
                # create a new instance by setting pk = None
                item.pk = None
```

File: product/gql_mutations.py [hunk 3]
```diff
@@ -353,6 +356,70 @@ def async_mutate(cls, user, **data):
...
+                item.product = new_product
+                item.save()
+
+        if duplicate_services:
+            new_product_services = ProductService.objects.filter(product=Product.objects.get(uuid=current_uuid,
+                                                                 validity_to__isnull=True))
+            for service in new_product_services:
+                service.pk = None
```
Comment (delcroip) [inline product/gql_mutations.py:416]:
```suggestion
                # create a new instance by setting pk = None
                service.pk = None
```

**Review by delcroip — APPROVED** [review]

**Pull request merged**